proto-grpc.md

```diff
@@ -0,0 +1,277 @@
+[bazelbuild/rules_proto](https://github.com/bazelbuild/rules_proto) provides the `proto_library`
+rule which can be used to establish libraries of `.proto` files and map their dependencies. For
+example, suppose we have the following dependency tree of `.proto` files:
+
+![Tree of proto files](./assets/proto-tree.png)
+
+For this set of files, we could define the following collection of `proto_library` targets. Note
+that some rulesets introduced later, such as stackb/rules_proto and rules_go, provide tooling
+through Gazelle for auto-generating these `proto_library` from the `.proto` files.
+
+```starlark
+load("@rules_proto//proto:defs.bzl", "proto_library")
+
+proto_library(
+  name = "proto_a",
+  srcs = ["A.proto"],
+)
+
+proto_library(
+  name = "proto_b",
+  srcs = ["B.proto"],
+)
+
+proto_library(
+  name = "proto_c",
+  srcs = ["C.proto"],
+  deps = [":proto_a", ":proto_b"], # Provides the depended files when calling protoc
+)
+```
+
+Using `bazel query`, we can view the graph that Bazel has assembled from the targets, showing it
+correctly reflects the dependency of `proto_c` on `proto_a` and `proto_b` (you must have
+DOT/GraphViz installed):
+
+`bazel query --output graph --nograph:factored 'kind(proto_library, //...)'  | dot -Tpng > graph.png`
+
+![Generated graph of proto_library targets](./assets/proto-bazel-query.png)
+
+
+Note that we could have grouped all of these `.proto` files under a single `proto_library` in this
+particular case. In real codebases these files may be under separate directories and managed by
+different groups or companies. The convention for splitting or grouping libraries varies by target
+language, but for most it is a matter of local preference. However, for some languages, files with
+the same internal [`package`](https://developers.google.com/protocol-buffers/docs/proto3#packages)
+specifier should normally be grouped into a single library. In general, you should be
+consistent with your grouping conventions within your workspace.
+
+At this point we have defined the abstract collections of `.proto` files, but these targets alone
+do not provide the source code artifacts to interact with the messages and services within our code,
+since we have not given protoc any plugins or target languages. Therefore, we need to introduce
+rulesets that can consume these targets and provide the language-specific plugins and knowledge to
+produce libraries.
+
+
+## The Ruleset Options
+
+Within the Bazel ecosystem, there are three main types of rulesets to consider when using Protobuf
+or gRPC:
+
+- **Canonical Dependencies**: The [Protobuf](https://github.com/protocolbuffers/protobuf/) and
+  [gRPC](https://github.com/grpc/grpc) source repositories contain a minimal selection of rules for
+  their key languages (C++, Python). These rules benefit from needing no further dependencies
+  beyond the basic repos. However, they are also mostly undocumented and not typically not well
+  supported.
```

Comment:
I'm not sure I'd ever recommend the rule sets from the protobuf repo. Not sure they even work with `proto_library`...

Reply:
Ok, I can dump that suggestion. Are the grpc rules in the grpc/grpc repo similarly odd?

Reply:
Answering my own question: grpc/grpc does appear to use rules_proto. I am going to leave the introduction to the 'Canonical' rulesets in place as context, but edit their introduction to dissuade their use. I will also remove them from the 'Picking a Ruleset' list.